chain-impl-mockchain/src/ledger/ledger.rs

```diff
@@ -723,66 +723,53 @@ impl Ledger {
-    pub fn apply_block(
+    /// This function must be called before applying fragments to the ledger. Do not use it if you
+    /// use `apply_block`.
+    pub fn apply_block_step1(
         &self,
-        ledger_params: &LedgerParameters,
-        contents: &Contents,
-        metadata: &HeaderContentEvalContext,
+        chain_length: ChainLength,
+        block_date: BlockDate,
     ) -> Result<Self, Error> {
         let mut new_ledger = self.clone();
 
         new_ledger.chain_length = self.chain_length.increase();
 
-        let (content_hash, content_size) = contents.compute_hash_size();
-
-        if content_size > ledger_params.block_content_max_size {
-            return Err(Error::InvalidContentSize {
-                actual: content_size,
-                max: ledger_params.block_content_max_size,
-            });
-        }
-
-        if content_hash != metadata.content_hash {
-            return Err(Error::InvalidContentHash {
-                actual: content_hash,
-                expected: metadata.content_hash,
-            });
-        }
-
         // Check if the metadata (date/heigth) check out compared to the current state
-        if metadata.chain_length != new_ledger.chain_length {
+        if chain_length != new_ledger.chain_length {
             return Err(Error::WrongChainLength {
-                actual: metadata.chain_length,
+                actual: chain_length,
                 expected: new_ledger.chain_length,
             });
         }
 
-        if metadata.block_date <= new_ledger.date {
+        if block_date <= new_ledger.date {
             return Err(Error::NonMonotonicDate {
-                block_date: metadata.block_date,
+                block_date,
                 chain_date: new_ledger.date,
             });
         }
 
         // double check that if we had an epoch transition, distribute_rewards has been called
-        if metadata.block_date.epoch > new_ledger.date.epoch && self.leaders_log.total() > 0 {
+        if block_date.epoch > new_ledger.date.epoch && self.leaders_log.total() > 0 {
             panic!("internal error: apply_block called after epoch transition, but distribute_rewards has not been called")
         }
 
         // Process Update proposals if needed
         let (updates, settings) = new_ledger.updates.process_proposals(
             new_ledger.settings,
             new_ledger.date,
-            metadata.block_date,
+            block_date,
         )?;
         new_ledger.updates = updates;
         new_ledger.settings = settings;
 
-        // Apply all the fragments
-        for content in contents.iter() {
-            new_ledger = new_ledger.apply_fragment(ledger_params, content, metadata.block_date)?;
-        }
+        Ok(new_ledger)
+    }
+
+    /// This function must be called after applying fragments to the ledger. Do not use it if you
+    /// use `apply_block`.
+    pub fn apply_block_step3(&self, metadata: &HeaderContentEvalContext) -> Self {
```

Comment:
so... step2 would be apply_fragments? Maybe `apply_block_finish` is better? Although that would suggest `start` instead of `step1`. The lack of `step2` threw me off a little at first.